common/djangoapps/student/models.py

```diff
@@ -1030,32 +1032,37 @@ def get_or_create_enrollment(cls, user, course_key):
         if user.id is None:
             user.save()
 
-        try:
-            enrollment, created = cls.objects.get_or_create(
-                user=user,
-                course_id=course_key,
-            )
+        # Even with MySQL set to use the READ COMMITTED isolation level, get_or_create sometimes
+        # raises an IntegrityError and the object in question doesn't appear in a subsequent get()
+        # call. Paired with non_atomic_requests, the retry logic here is meant to prevent IntegrityErrors
+        # from causing authenticated requests to fail.
+        retry = 0
+        while True:
```

Comment:
@schenedx the while loop and conditional you commented on below are used so that we don't sleep any longer than is required. If you use `while retry <= MAX_RETRIES:` without the conditional and actually end up retrying `MAX_RETRIES` times, you'll end up sleeping after the last retry fails, before raising an exception. That's undesired.